Source/WebCore/dom/InternalObserverFirst.cpp

```diff
@@ -0,0 +1,113 @@
+#include "InternalObserverFirst.h"
+
+#include "AbortController.h"
+#include "AbortSignal.h"
+#include "Exception.h"
+#include "ExceptionCode.h"
+#include "InternalObserver.h"
+#include "JSDOMPromiseDeferred.h"
+#include "Observable.h"
+#include "ScriptExecutionContext.h"
+#include "SubscribeOptions.h"
+#include "Subscriber.h"
+#include "SubscriberCallback.h"
+#include <JavaScriptCore/JSCJSValueInlines.h>
+
+namespace WebCore {
+
+class InternalObserverFirst final : public InternalObserver {
+public:
+    static Ref<InternalObserverFirst> create(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
+    {
+        Ref internalObserver = adoptRef(*new InternalObserverFirst(context, signal, WTFMove(promise)));
+        internalObserver->suspendIfNeeded();
+        return internalObserver;
+    }
+
+private:
+    void next(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->resolve<IDLAny>(value);
+        (Ref { m_signal })->signalAbort(JSC::jsUndefined());
+    }
+
+    void error(JSC::JSValue value) final
+    {
+        (Ref { m_promise })->reject<IDLAny>(value);
+    }
+
+    void complete() final
+    {
+        InternalObserver::complete();
+        (Ref { m_promise })->reject(Exception { ExceptionCode::RangeError, "No values in Observable"_s });
+    }
+
+    void visitAdditionalChildren(JSC::AbstractSlotVisitor&) const final
+    {
+    }
+
+    void visitAdditionalChildren(JSC::SlotVisitor&) const final
+    {
+    }
+
+    InternalObserverFirst(ScriptExecutionContext& context, Ref<AbortSignal> signal, Ref<DeferredPromise>&& promise)
+        : InternalObserver(context)
+        , m_signal(signal)
+        , m_promise(WTFMove(promise))
+    {
+    }
+
+    Ref<AbortSignal> m_signal;
+    Ref<DeferredPromise> m_promise;
+};
+
+void createInternalObserverOperatorFirst(ScriptExecutionContext& context, Ref<Observable> observable, SubscribeOptions options, Ref<DeferredPromise>&& promise)
+{
+    Ref signal = AbortSignal::create(&context);
+    Vector<Ref<AbortSignal>> dependentSignals = { WTFMove(signal) };
+
+    if (UNLIKELY(options.signal))
+        dependentSignals.append(*options.signal);
+    auto dependentSignal = AbortSignal::any(context, dependentSignals);
+
+    if (UNLIKELY(dependentSignal->aborted()))
+        return promise->reject<IDLAny>(dependentSignal->reason().getValue());
+
+    dependentSignal->addAlgorithm([promise](JSC::JSValue reason) {
+        promise->reject<IDLAny>(reason);
+    });
+
+    // FIXME: We really should be passing `signal` down instead of `dependentSignal`. As per point 3 here https://wicg.github.io/observable/#dom-observable-first
+    //    but we have this failing test as a result: https://github.com/WebKit/WebKit/blob/2d7df09cf3f08a332003e47dd0664f4c3423604a/LayoutTests/imported/w3c/web-platform-tests/dom/observable/tentative/observable-first.any.js#L66
+    auto observer = InternalObserverFirst::create(context, dependentSignal, WTFMove(promise));
```

Comment:
@darinadler @cdumez Id like to focus your attention this line specifically in relation to this previous comment: https://github.com/WebKit/WebKit/pull/35261#discussion_r1804687496

Reply:
Does the test fail because the test is incorrect?

Reply:
Yes, for what it's worth, I agree that we should be passing `signal` here, not `dependentSignal`, per the spec.

Reply:
@cdumez I have change this line to read _as we'd expect_.